### ISSUE TITLE
Improve test coverage against multiple DBMSes

### ISSUE THREAD

**Issue by rhauch** [confluentinc/kafka-connect-jdbc#342]
Currently the Maven build runs JUnit integration tests using an embedded Derby database, and does not run the integration tests against any other DBMSes. This makes it very difficult to test whether the connector works against different DBMSes and hampers the ease with which we verify DBMS-specific logic. 

We should change the Maven build to enable running the integration tests against different DBMSes. We don't want to be careful to not increase the build time for normal development, and so we should consider separate profiles for each DBMS that are not enabled by default. Our Jenkins build can then run all of the profiles. We would likely need to define additional integration tests that connect, create the necessary tables and data, run the source/sink connector, verify the results, and tear down the database changes.

We also don't want to require nor rely upon developers having the relevant DBMSes installed locally just to run the tests. First of all, they may already have different versions than that which we want to test. Second, they may not have them installed in the first place. Third, we don't want to interfere with any existing databases in those systems. In summary, the build should be self-contained. 

One very attractive option is to use Docker containers. The Maven build's profile for a DBMS could specify the Docker image as well as DBMS-specific system properties that might be required as inputs to the tests (e.g., connection urls, usernames, passwords, etc.). Of course, we'd be limited to the DBMSes that have publicly-available Docker images, but that gets us PostgreSQL, MySQL, Sybase, DB2, SQL Server, and others. And, as long as Docker were installed, the build for these DBMSes would be completely automated. Some proprietary DBMS vendors make it easy to build your own Docker images with their DBMS, but those images cannot be shared -- we could accommodate that by providing instructions that developers could use to build the images locally before they run the build with the respective profile.

There are other options as well, but they have pros and cons in terms of the prerequisites.

Please weigh in with your thoughts and opinions. This is a non-trivial issue, so it'd be great to split up the work and have lots of people helping out.

**Comment by ciaranarcher** [comment]
One approach worthy of investigation is to use [Docker Compose](https://docs.docker.com/compose/overview/). 

Similar to the approach taken in this [article](https://hharnisc.github.io/2016/06/19/integration-testing-with-docker-compose.html) it might be possible to create set of sub-directories, one per database, each of which contains a `docker-compose.yml` file. The directory would also contain any drivers for the database (e.g. the MySQL JDBC driver etc.)

The test runner (perhaps just a `Makefile`) in the parent directly could then, for each database under test: 

* Run `docker compose up` in the relevant directory. 
* Run the common set of tests, injecting the correct JDBC URL and linking the correct database driver jars. 
* Repeat for all sub-directories and thus all supported databases. 

A simple (and default) `make test` would just run the usual tests with Derby. 

I could have a stab at a POC of this, if folks thought it wasn't crazy and might actually work.

**Comment by rhauch** [comment]
Yes, just requiring the developer to run the correct Docker command prior to the tests would work. But I have a couple of questions.

First, you're essentially using Docker Compose to easily start a single container, right? 

Second, why not use existing Maven plugins to automatically start Docker containers from within Maven as part of the integration test phase of Maven, based upon the Maven profile. That would be much easier from a developer perspective: just run `mvn clean install -Pmysql`, `mvn clean install -Ppostgres`, `mvn clean install -Pmysql,postgres,sybase`, or just `mvn clean install` to skip the integration tests altogether. The [Debezium project](http://debezium.io) does this ([code here](https://github.com/debezium/debezium)).

I'd want to avoid adding a different build system (e.g., make) to the project that is already using Maven.

**Comment by kkonstantine** [comment]
Leveraging Docker for integration tests with different databases seems like a very natural fit. I'm pretty much in agreement with this suggestion. 

Having said that, I also think we should not diverge from the current build process. Which means we should keep using maven and leverage its plugins. The way @rhauch is describing the parameters makes a lot of sense, probably adding yet another one such as `-PdbAll` to run all tests would this even better (my inspiration with names and acronyms is pretty flaky - but my point is that a flag that instructs mvn to run all db-based integration tests should be provided).

**Comment by ciaranarcher** [comment]
@rhauch @kkonstantine 

> First, you're essentially using Docker Compose to easily start a single container, right?

I'm proposing that each integration test sub-directory would include the `cp-kafka-connect` Docker image (and any dependencies) and the database image you wish to test against. The build step would ensure that: 
* we compile the latest JDBC connector JAR based on the current code (e.g. on the branch the engineer is working on) 
* and Kafka Connect uses that JAR (i.e. overwrite the bundled version in the Docker image)

As opposed to running a set of tests similar to what is in place now, I'm talking about a set of real integration tests that would write some data database under test, and then consume and verify those messages ended up in Kafka. So I guess maybe a little higher level than @rhauch was proposing. 

Regarding Maven usage: 

I totally understand keeping any complexity this approach adds behind Maven. My suggestion probably comes from the fact I'm not a Java developer by profession (usually Go, Ruby) so I'll defer to you both there.

**Comment by rhauch** [comment]
@ciaranarcher IIUC, using the `cp-kafka-connect` docker image to run Connect would also require running Kafka containers, plus we'd need to run a container for each DBMS. I agree that such _system tests_ would be great to have, and here using Docker Compose is probably the most straightforward way to do this.

But I'd like to first focus on _integration tests_ to allow us to test the basic functionality of the connector with each DBMS without having to run Connect, Kafka, Zookeeper, the DBMS, and finally deploy the connector. IOW, I'd like to still use JUnit to write tests that verify component behavior against the real DBMS. For example: is the connector able to properly read the database metadata and construct the proper `Schema` object for a table; does the dialect generate a valid `CREATE TABLE` statement for the many different combinations of field schemas; etc.
